frontend/components/forms/DynamicForm.tsx

```diff
@@ -0,0 +1,226 @@
+import axiosWithoutAuth from "@/axiosWithoutAuth";
+import InstantSearch from "@/components/search/InstantSearch";
+import {
+  Box,
+  Button,
+  Divider,
+  FormControl,
+  Grid,
+  InputLabel,
+  MenuItem,
+  OutlinedInput,
+  Select,
+  Typography,
+} from "@mui/material";
+import Card from "@mui/material/Card";
+import TextField from "@mui/material/TextField";
+import { useRouter } from "next/router";
+import { ParsedUrlQueryInput } from "querystring";
+import { FC, MutableRefObject, useContext, useEffect, useRef, useState } from "react";
+import PageTransitionContext from "../PageTransitionContext";
+
+const DynamicForm: FC = () => {
+  const [type, setType] = useState("");
+
+  const handleFormType = (event: { target: { value: string } }) => {
+    setType(event.target.value);
+  };
+
+  return (
+    <Card elevation={5}>
+      <FormControl sx={{ minWidth: "20rem", m: "1rem" }}>
+        <InputLabel id="select-form-type">Type of Content</InputLabel>
+        <Select
+          labelId="select-form-type"
+          id="select-form"
+          value={type}
+          label="Type of Content"
+          onChange={handleFormType}
+        >
+          <MenuItem value={"none"}>
+            <em>None</em>
+          </MenuItem>
+          <MenuItem value={"quotes"}>Quote</MenuItem>
+          <MenuItem value={"entities"}>Entity</MenuItem>
+          <MenuItem value={"images"}>Image</MenuItem>
+          <MenuItem value={"occurrences"}>Occurrence</MenuItem>
+          <MenuItem value={"topics"}>Topic</MenuItem>
+        </Select>
+      </FormControl>
+      {type && type != "none" && (
+        <>
+          <Divider sx={{ borderBottomWidth: 2 }} />
+          <Box
+            sx={{
+              flex: "1 1",
+              display: "flex",
+              alignItems: "center",
+              justifyContent: "center",
+              margin: "1.5rem 1rem 1.5rem 1rem",
+            }}
+          >
+            <Box>
+              <DynamicFormFields type={type} />
+            </Box>
+          </Box>
+        </>
+      )}
+    </Card>
+  );
+};
+
+interface DynamicFormProps {
+  type: string;
+}
+
+const fields = ["topics"] as const;
+type Field = typeof fields[number];
+type FieldsRef = MutableRefObject<Record<Field, any>>;
+type FieldCallbacks = Record<Field, (value: ParsedUrlQueryInput[string]) => void>;
+
+//Dynamic fields form
+//Can be called using the DynamicForm component or by using the DynamicFormFields prop and specifying the 'type' prop
+const DynamicFormFields: FC<DynamicFormProps> = ({ type }: DynamicFormProps) => {
+  const [formData, setFormData] = useState([]);
+  const [choiceValue, setChoiceValue] = useState<string[]>([]);
```

Comment:
It looks like the same choice state values here are used for all `Select` fields rendered, so it breaks when more than one value is selected.

![image](https://user-images.githubusercontent.com/24942155/140678709-b751e0ec-060b-481e-b9d0-b5c25553fb11.png)

Reply:
Hmm okay. Would I need to create a state value for each ManyRelatedField that is displayed before hand? Or is there an easier way of doing this?

Reply:
https://github.com/ModularHistory/modularhistory/blob/e9ee643744f370a0c4a4fcdd1b1de8f02d295782/frontend/components/forms/DynamicForm.tsx#L99-L102

The way this code works is to have a single object `fieldsRef` holding values for each field, and have unique callbacks for each field so that each write to that object with a unique key. You might just be able to pass the correct callback `fieldCallbacks.current[field.name]` to the `Select` components, but it's also possible the generated callback would need to be modified to handle `Select`'s `onChange` arguments ([docs](https://mui.com/api/select/#props)). If that's true, then you could just add a field type check to L100-102, something like
```jsx
fields.map(({name, type}) => (
  [
    name, 
    type === "M2M" 
      ? (whatever, args, value) => {fieldsRef.current[name] = value} 
      : (value) => {fieldsRef.current[name] = value} 
  ]
))
```

It looks like you're hardcoding the fields right now instead of pulling names from the API response, so that will need to be fixed. We can only do that on the search form since the fields are hardcoded.
https://github.com/ModularHistory/modularhistory/blob/e9ee643744f370a0c4a4fcdd1b1de8f02d295782/frontend/components/forms/DynamicForm.tsx#L75-L76

Reply:
Currently, in this implementation, Jacob said that it would be a good idea to just copy over the topics search from the SearchForm.tsx. This is currently the only field name that is hardcoded as that topic field use the topics search instead of creating its own field.

This isn't the best as it is hardcoded but currently, all types have a topic field in the API response.

Reply:
As far as I can tell, utilizing the pre-existing topics field is handled entirely by the logic here.
https://github.com/ModularHistory/modularhistory/blob/e9ee643744f370a0c4a4fcdd1b1de8f02d295782/frontend/components/forms/DynamicForm.tsx#L130-L132
It doesn't look like it depends on having this array hard-coded.